### PR TITLE
add retrain logic for insufficient data

### PR DESCRIPTION
**Pull request by s0nicboOm**
Tackling retraining for insufficient data.

image

image

### CODE REVIEW

File: numalogic/udfs/tools.py [hunk 1]
```diff
@@ -164,43 +164,101 @@ def __init__(self, r_client: redis_client_t):
...
         return f"TRAIN::{':'.join(keys)}"
 
-    def __fetch_ts(self, key: str) -> tuple[Optional[str], Optional[str]]:
+    def __fetch_ts(self, key: str) -> tuple[Optional[str], Optional[str], Optional[str]]:
         try:
             data = self.client.hgetall(key)
         except RedisError:
             _LOGGER.exception("Problem  fetching ts information for the key: %s", key)
-            return None, None
+            return None, None, None
```
Comment (ab93) [inline numalogic/udfs/tools.py:172]: convert to named tuple

**Review by ab93 — CHANGES_REQUESTED** [review]
Minor comment

File: numalogic/config/_config.py [hunk 2]
```diff
@@ -90,6 +90,7 @@ class TrainerConf:
     retrain_freq_hr: int = 24
     retry_sec: int = 600  # 10 min
     batch_size: int = 64
+    data_granularity_sec: int = 60
```
[the remark this answers is not in the document]
Comment (s0nicboOm) [reply numalogic/config/_config.py]: sure

File: numalogic/udfs/tools.py (hunk 1 above)
Comment (s0nicboOm) [reply numalogic/udfs/tools.py]: sure

File: numalogic/config/_config.py [hunk 3]
```diff
@@ -90,6 +90,7 @@ class TrainerConf:
     retrain_freq_hr: int = 24
     retry_sec: int = 600  # 10 min
     batch_size: int = 64
+    data_freq_sec: int = 60
```
Comment (ab93) [inline numalogic/config/_config.py:93]: Let's add a docstring for this class

File: numalogic/udfs/tools.py [hunk 4]
```diff
@@ -17,6 +17,14 @@
 _LOGGER = logging.getLogger(__name__)
 
 
+class _DedupMetadata(NamedTuple):
+    """namedtuple for Dedup Metadata."""
```
Comment (ab93) [inline numalogic/udfs/tools.py:21]:
```suggestion
    """Data structure for Dedup Metadata."""
```

**Review by ab93 — APPROVED** [review]
Minor comments. Otherwise LGTM

**Pull request merged**